### PR TITLE
Upgrade version numberring to X.Y.<build>

### PR DESCRIPTION
**Pull request by simon33-2**
## Overview
The discussed change to version numbering.

## Functional Changes
Removes unnecessary double zero in the version numbering. This also allows Triple-A to know what build it was and if something is older or newer.

## Manual Testing Performed
- ..
Games are able to be started, saved and loaded. Can't load old version games.

### CODE REVIEW

**Comment by ssoloff** [comment]
Before getting too deep into the review, my vote would be to keep the build ID separate from the semantic portion of the version: that is, _major.minor.point.buildId_ (preferably with _buildId_ as a string and all other components as integers).  If we eventually adopt semver, it would be awkward to have the _point_ component change simply because a new build was started.  There's all sorts of things that could trigger a new build without having an actual code change (e.g. creating a new branch). A new build of the same code shouldn't express a semantic change in the deliverable, although there should be a component that identifies it as a unique build (because things like the build environment could have changed between builds).

Also, I'd like to keep _buildId_ as a string so we don't have to deal with all the hooeyness of artificially converting `dev` to an integer.  That was mostly a hack to maintain compatibility with the existing format.  Since we can break compatibility, we might as well make it `<int>.<int>.<int>.<string>`.

Tagging @DanVanAtta, @RoiEXLab, and @ron-murhammer for their thoughts on what they'd like to say.  I'm okay with proceeding as-is if that's what the team decides.

**Comment by RoiEXLab** [comment]
@ssoloff I personally would like to see a version format that aligns with the semver spec, where the build number is part of the "build metadata": https://semver.org/#spec-item-10
TL;DR major.minor.patch+buildNumberAndPotentialDistributionType

**Comment by ssoloff** [comment]
@RoiEXLab :+1: Standards are always good.  It's basically what I was thinking of, just with a different delimiter.

I'd be curious if someone already wrote a semver-compatible `Version` type for the JVM.  (Oh, look at that: https://github.com/zafarkhaja/jsemver :smile:)

**Comment by simon33-2** [comment]
Perhaps a fair point on making the buildId a string although then how are we going to do comparisons? We still want to be able to compare newer and older don't we?

What's the point of maintaining the third number?

Regarding going to semver, haven't we just decided we aren't ready to do that when we said we wanted the breaking release to be 1.10?

**Comment by ssoloff** [comment]
@simon33-2 

> Perhaps a fair point on making the buildId a string although then how are we going to do comparisons? We still want to be able to compare newer and older don't we?

Sorry for not clarifying that point.  The build ID should be ignored during comparisons because it has no semantic meaning.  It's simply there to give devs a cross-reference to the build that produced the artifact.

I admit I don't understand our legacy _major.minor.point.micro_ convention and why we never bumped even _micro_ during stable releases.  I realize there's some "compatibility" reason buried in there, but, IMO, it was always confusing.

> What's the point of maintaining the third number?

Because hopefully we'll adopt semver one day. :smile:  ...and actually bump the `patch` component when we promote a new stable release that simply contains bug fixes.

> Regarding going to semver, haven't we just decided we aren't ready to do that when we said we wanted the breaking release to be 1.10?

Yes, the team agreed not to bump the major version for the new incompatible release.  As I said above, I'm still hopeful that we'll formally adopt semver in the future. :crossed_fingers:  Even if we don't embrace semver now, following a standard format at least means one less thing to document.

**Comment by simon33-2** [comment]
>>     What's the point of maintaining the third number?

>Because hopefully we'll adopt semver one day. smile ...and actually bump the patch component when we promote a new stable release that simply contains bug fixes.

Sorry but that argument is ridiculous. Maintain it to remove it later?

**Comment by ssoloff** [comment]
> Sorry but that argument is ridiculous. Maintain it to remove it later?

Sorry if that's the way it came across.  I'm not suggesting removing the patch component at a later time.  I'm suggesting we simply use it for what it was designed for.

Thinking about this a bit more, if we choose to never adopt semver, and continue to follow the release process we've been using (nominate a build for stable, wait for QA approval, and then promote it to stable), we really only need a single semantic version component.  That is _major.buildId_, where _major_ is simply used to indicate compatibility.  I still think _buildId_ wouldn't need to be involved in the comparison in that case, but I could be missing a scenario where that would be required.

**Comment by ssoloff** [comment]
@DanVanAtta @ron-murhammer Would appreciate your opinions here.  We probably want to get this right, at least for the 1.10 release, so that we're not flip-flopping between different formats/implementations and confusing everybody.

The main questions are

1. We've decided to not adopt semver for the 1.10 release, but do we plan on adopting it in the future?
1. Should the build ID be ignored when comparing one version to another?

**Comment by veqryn** [comment]
I'll be honest, I don't think SemVer makes sense for this project.

If you had a library or an api, that was referenced and used by other coders/projects, then yes it would make sense.

But you don't. TripleA does not have an API.

What constitutes a "breaking" change to TripleA?  I'm pretty sure if someone was trying to use our deployed jar files, they've broken many many times by now.

Instead, you all should look at what other executable client software does for versioning.

You could target release date and go with a year + month version, like ubuntu or windows.

You could go with feature releases and bug fix releases, where the feature releases bump major version numbers and the bug fix releases are the minor numbers.

Pretty much anything other than semver would be more applicable to triplea.

**Comment by simon33-2** [comment]
@veqryn I think the point is that 1.9 is in theory compatible with all other versions of 1.9. There were a few breaks in this theory but I'm not aware of any from build 3010 onwards, at least OTOH.

It is incompatible with all versions of 1.8 and all previous versions.

The most annoying thing about 1.9's versioning is that it didn't report the **** build Id. Why the hell not? Please don't do this to users and people trying to help them in later versions.

**Comment by RoiEXLab** [comment]
@veqryn I disagree on this.
While it's true that triplea isn't an API as a jar, it's very much an API for map makers.
Having semver strings could be a very useful way to define map xml compatibility, whenever a minor map xml functionality gets added the patch number would get incremented and the engine could reliably check that, because the build number is currently just an "aesthetic" thing.

**Comment by veqryn** [comment]
So, I think that with TripleA should maintain a promise of backwards compatibility for any and all maps made today.

Given that a breaking change will never, or almost never, be done (if it can even be defined), our major version number will never change.

I feel like we should adopt something like android's version number scheme.
The major version number should be incremented whenever a major set of features are released.  The minor version number for small sets of features and bug fixes.  The patch number for our build numbers.

If we have breaking changes, they can be deprecated on a schedule: backwards compatibility will be maintained for 2 major version numbers or 6 months, whichever comes first (or something like that).

Summary: I think it is better to use to use TripleA's major and minor numbers to communicate new features directly to the end user.

**Comment by simon33-2** [comment]
I'm just pointing out that your comment was factually incorrect.

**Comment by DanVanAtta** [comment]
@simon33-2 I thumbed down your last comment as it was not helpful for moving the conversation forward. Basically you ended with saying "your're wrong". That is not helpful on multiple levels.

TripleA is a "loud" community, perhaps one of its biggest weaknesses. We can work against each other and let TripleA flounder, or we can work together. Please know, whenever we work against each other, anyone not for TripleA is laughing.

Please ensure that your comments are always constructive and move the conversation forward. In this case please instead you could have said something like
- "I disagree with the statement '`quote`' for the following reasons: ...."

Reading the above, I'm left confused what could have been factually wrong by a statement as qualified as "almost never", which makes me wonder if I even understand what you are disagreeing with @simon33-2 

I think we can agree to let this be 'water under the bridge' for now, let's remember we are all working for the betterment of TripleA, and the more effective we are the more the community would win.

**Comment by DanVanAtta** [comment]
I think everyone is right here, but we are not addressing the full complexity (and audience) around version numbers. Building on this idea:

> While it's true that triplea isn't an API as a jar, it's very much an API for map makers.

It's also an API for network compatibility, and similar for save game compatibility. Thus in part what complicates the version numbering question is that we have "major" version compatibility along *multiple* dimensions. It's not just "will this code compile and execute against this other version", but it's a significant difference between network compatibility and map compatibility. This is simply not captured in a single semver.

The other issue, where I think @veqryn is right to point out that semver does not make sense for TripleA is that a semver number is not useful/appropriate for our users. Thinking of a recent example a few weeks ago in lobby, asking someone if they were on a recent version, they said yes, and they were on version '*3635'. 

I think the way to look at it is that TripleA actually needs *4* version numbers:
- a release or product number that is user-facing (eg: "18.11.<build_id>") 
- dev-facing network compatibility
- dev-facing save-game compatibility
- dev-facing map-parsing compatibility

Having a compatibility  number for save-game and map-parsing would give us a number of benefits, it would let us support both new and old features easily, letting us not break backward compatibility while still able to add new non-compatible features.

**Comment by DanVanAtta** [comment]
I'll create an issue to discuss version number in more detail. We need to align I think on a team of how we will do this at an architectural/system level before we can really make any reasonable code changes. @simon33-2 I think that means we will need to revisit these code updates. In part it's pretty clear I think that is the case as our conversation has very little to do with the specific diff presented here.

**Comment by simon33-2** [comment]
@ssoloff You don't feel that having a meaningful message to users is sufficient reason to retain the "m_" prefixes in this case?

**Comment by ssoloff** [comment]
@simon33-2 I'm missing some context here.  Are you referring to [this discussion](https://github.com/triplea-game/triplea/pull/4332#issuecomment-437095971)?

**Comment by simon33-2** [comment]
Yes I am. Sorry about that confusion.

**Comment by simon33-2** [comment]
@DanVanAtta I found veqryn's comment that we would never or almost never make an incompatible release strange, particularly given that we are currently working on one. I think that 1.9 has lasted a lot longer than earlier versions. Do we really need 3 numbers denoting compatibility?

Anyway, perhaps I could have worded my now deleted comment better so I'm sorry for that.

**Comment by simon33-2** [comment]
> Thinking of a recent example a few weeks ago in lobby, asking someone if they were on a recent version, they said yes, and they were on version '*3635'.

This is exactly the reason why we need to be able to do comparisons on the build ID. As far as I'm aware, we resolved all the issues with moving to 1.9.0.1 or more during the development of 1.9 but at no point did we ever change the last version number. I can't see anything being different in a later version. Is anyone prepared to argue differently?

**Comment by DanVanAtta** [comment]
> @DanVanAtta I found veqryn's comment that we would never or almost never make an incompatible release strange, particularly given that we are currently working on one. 

Incompatible release typically means breaking savegame compatibility, and *usually* network compatibility is often updated, but not necessarily. Map compatibility changing was an exception going from 1.8 to 1.9, and was because we migrated where maps were stored to begin with. Typically map compatibility is *never* broken between any releases. 1.9 was an exception as we went from SVN to github hosting and games needed new maps regardless. 

When designing how maps were set up I attempted to get it so would have an option to update maps. That was a mistake as the proper answer is to better support immutable maps and to ensure we can sanely have new games always read/parse old games. Furthermore given the pain that has been largely felt when updating maps so that new versions can no longer read old, it's something we should probably never repeat. I strongly agree we should then strive to ensure we can always read all existing maps. Whether that translates to an "incompatible release" is semantics. I hope we can shift the technology of TripleA to no longer have our hands tied for compatibility reasons and allow for easier backward compatibility between new versions to old, so this would be much more of a non-issue.

I really ideally want us to beef up the testing and then undo our 'latest prelease' override so that any time after we merged we'd build and release a brand new release version on the spot. It's quite the road to get there though.


> I think that 1.9 has lasted a lot longer than earlier versions. Do we really need 3 numbers denoting compatibility?

Let's move the conversation of how to do version numbering to an issue. PR's should not be for discussing relatively high-level/strategic initiatives. It's been some time since we've discussed any of the actual code changes. Even if things were really well tested here, it's not clear that simply lopping off a version number is the right thing to do. Why not just lop off 2, or even 3? But again, that conversation is greater than this single PR and I think should move entirely to: https://github.com/triplea-game/triplea/issues/4346

@simon33-2 what is your plan forward on this PR? I don't think we can seriously review it until we've decided where to go on version numbering. I'd like a plan instead of very haphazard improvement. Is the first step in that plan to at least expose the build number in the UI? That seems to be about where we are going here, but there is a lot of testing missing. If on the other hand this is not the right first step at all, then we'll waste effort. 

If the plan is to first expose in the UI the full version number so we no longer have a silent 'build version' at the end of our 5 digit number, that is okay. If so, there is pretty extensive testing that should happen before this can be merged (that falls to PR author @simon33-2). Any discussion on long-term objective with version number does not belong in a PR.

**Comment by simon33-2** [comment]
Frankly, I was expecting an answer to my post in the issue you opened.

If I had have suggested removing only the 4th number I'm 75% sure someone would have said that we can't do it because we aren't removing the third number. I was just trying to avoid the utter crap in 1.9 where we didn't log the build number. Not acceptable.

**Comment by simon33-2** [comment]
One more thing, I know you love the idea of having no more incompatible releases but is that really an achievable goal? There is always the possibility of a feature (or bug fix) being added requiring more information to be held in the save game which will break serialisation won't it? I'm not sure what the fix to this might be, if there even is one possible.

**Comment by DanVanAtta** [comment]
> Frankly, I was expecting an answer to my post in the issue you opened.

The response is not necessarily trivial.

@simon33-2 could you please answer this question:

> what is your plan forward on this PR? 

Do you need it to  be just reviewed? It looks like there is a lot of testing to be done, inspecting the code I'm not sure if all of these changes would work straight-up. Is your plan to confirm that we are okay with dropping a number so that the build number is no longer hidden, and then test this PR, and then let us know when it is ready for a review? Do you have a differnet plan.

**Comment by simon33-2** [comment]
Yeah, I was thinking that this should be reviewed. 

I was under the impression that there was general agreement that the 3rd number was useless. Perhaps I'm incorrect on that point if you're asking the question?

**Comment by simon33-2** [comment]
Well the travis build succeeded. Not entirely sure what "build pipeline" refers to.

**Comment by DanVanAtta** [comment]
@simon33-2 this looks pretty complete for the intended update - thank you. Focusing on the goal of not having hidden build numbers, this moves us a good step toward that by simply lopping off a number. I think we all agree the build ID should not be the *5th* version number.

@RoiEXLab you voiced objection, let's please continue the discussion on how to handle versioning in the long-term in issue: https://github.com/triplea-game/triplea/issues/4346 ; If we are looking to move to semver, if anything making the 5th build number a 4th number automates the update to the micro number and moves us closer to that goal. In #4346 there is a debate on how to handle internal modules, whether those would have version numbers themselvles (potentially semver), and if the engine should have a semver, and if that semver ought to public facing or internal-only with an external release name.

**Comment by DanVanAtta** [comment]
@simon33-2   Wait, I was wrong to say one number is lopped off - there are two removed here. This is actually going to cause problems as the 'point' number becomes the build number. I'll comment in-code where I think we need to make an update.

File: game-core/src/main/java/games/strategy/engine/GameEngineVersion.java [hunk 1]
```diff
@@ -34,7 +34,7 @@ public static GameEngineVersion of(final Version version) {
   public boolean isCompatibleWithEngineVersion(final Version engineVersion) {
     checkNotNull(engineVersion);
 
-    return version.withMicro(0).equals(engineVersion.withMicro(0));
+    return version.withPoint(0).equals(engineVersion.withPoint(0));
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/engine/GameEngineVersion.java]: With 3 numbers, and build number as the 3rd, it will be the 'point' version. I think this should then be 'withMinor', and you can remove the 'point' variable from Version.java.

build number was added in a way that the code was totally unaware of, it's happy splitting on periods and ignoring extra numbers. I don't think we want engine compat to depend on any change, 'minor' is actually pretty in-line with how we've been using it.

Comment (simon33-2) [reply game-core/src/main/java/games/strategy/engine/GameEngineVersion.java]: Regarding your comment:
> This is actually going to cause problems as the 'point' number becomes the build number

I don't think it will cause any problems as I have coded it. I'm pretty sure I've tested all reasonable scenarios, although it was over a month now.

**Comment by ssoloff** [comment]
@simon33-2 One other thing to consider is how _lobby_server.yaml_ is parsed.  This file is processed by both 1.9 and 1.10 clients when they connect to the lobby.  It currently contains entries for both 1.9 and 1.10.  I _think_ it will still work with this change as that file exists today, but the question is will leaving an entry for `1.10.0.0` (instead of `1.10.0`) screw us up going forward?  Can it be changed to `1.10.0` in this PR without a 1.9 client getting confused?

**Comment by simon33-2** [comment]
I believe that the lobby_server.yaml should still work for 1.9 with or without change. If it's parsed by Version.java, there is no difference between a number that is missing or a number that is 0. However, there could be an issue for 1.10. If the lobby_server.yaml says "1.10.0" or "1.10" and the client version is 1.10.12345, then that might be deemed incompatible if it's parsed by some other processor.

I've added a commit to address ssoloff's concerns but there might still be a need to check that the lobby still works. I've tried starting the lobby with "./gradlew :lobby:run" and connecting to it with "TripleA -Ptriplea.lobby.host=192.168.0.23" but it seems that it connects to the normal lobby. Is there something else I need to do to get the connect to go to my own lobby?

File: config/triplea/schemas/lobby_server.json [hunk 2]
```diff
@@ -10,7 +10,7 @@
     "properties": {
       "version": {
         "description": "The lobby server engine version",
-        "type": "string",
+        "type": [ "string", "number" ],
```
Comment (ssoloff) [inline config/triplea/schemas/lobby_server.json]: Why not just change the regex to something like `^\\d+(\\.\\d+){0,3}$`?  A `number` will allow  stuff like `-1.10` through the validator.

**Comment by ssoloff** [comment]
> I've added a commit to address ssoloff's concerns but there might still be a need to check that the lobby still works. I've tried starting the lobby with "./gradlew :lobby:run" and connecting to it with "TripleA -Ptriplea.lobby.host=192.168.0.23" but it seems that it connects to the normal lobby. Is there something else I need to do to get the connect to go to my own lobby?

@simon33-2 Unfortunately, connecting to your own lobby won't help testing a custom _lobby_server.yaml_.  No matter if you connect to the production lobby or your own lobby, _lobby_server.yaml_ is retrieved by the `LobbyServerPropertiesFetcher` class in the client.  That class is currently hard-coded to always fetch the YAML file from the `master` branch of this repo.  We used to have a way to override the URL of this file so you could redirect it to your local file system.  However, I can no longer find that feature.

So, to test your local file, you'll have to hack up the `LobbyServerPropertiesFetcher#downloadAndParseRemoteFile()` method.  For example:

```diff
diff --git a/game-core/src/main/java/games/strategy/engine/config/client/LobbyServerPropertiesFetcher.java b/game-core/src/main/java/games/strategy/engine/config/client/LobbyServerPropertiesFetcher.java
index fe7007932..2b96dc27c 100644
--- a/game-core/src/main/java/games/strategy/engine/config/client/LobbyServerPropertiesFetcher.java
+++ b/game-core/src/main/java/games/strategy/engine/config/client/LobbyServerPropertiesFetcher.java
@@ -122,18 +122,18 @@ public final class LobbyServerPropertiesFetcher {
       final Version currentVersion,
       final BiFunction<File, Version, LobbyServerProperties> propertyParser)
       throws IOException {
-    final DownloadUtils.FileDownloadResult fileDownloadResult = fileDownloader.download(lobbyPropFileUrl);
-
-    if (!fileDownloadResult.wasSuccess) {
-      throw new IOException("Failed to download: " + lobbyPropFileUrl);
-    }
+//    final DownloadUtils.FileDownloadResult fileDownloadResult = fileDownloader.download(lobbyPropFileUrl);
+//
+//    if (!fileDownloadResult.wasSuccess) {
+//      throw new IOException("Failed to download: " + lobbyPropFileUrl);
+//    }
 
     final LobbyServerProperties properties =
-        propertyParser.apply(fileDownloadResult.downloadedFile, currentVersion);
+        propertyParser.apply(new File("/home/simon/triplea/lobby_server.yaml"), currentVersion);
 
     // delete file after it has been used. If there there was an IOException, the 'deleteOnExit' should
     // kick in and delete the file. (hence there is no try/catch/finally block here)
-    fileDownloadResult.downloadedFile.delete();
+//    fileDownloadResult.downloadedFile.delete();
     return properties;
   }
 }
```

<hr>

BTW, for future reference, to connect to your own lobby, go into **Engine Preferences > Testing** and change the **Lobby Host Override** and **Lobby Port Override** fields to **localhost** and **3304**, respectively (assuming you're running the lobby on the same machine and you didn't change the default port).

File: game-core/src/main/java/games/strategy/util/Version.java [hunk 3]
```diff
@@ -35,17 +34,9 @@ public Version(final int major, final int minor) {
    * Constructs a Version object without the micro version, defaults to 0.
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/util/Version.java]: This Javadoc is no longer applicable.  The one below that was removed would probably be more appropriate.

There's a few other comments within this file that still make reference to the four-component form.  Those should probably be updated to reflect the current state.

File: game-core/src/main/java/games/strategy/util/Version.java [hunk 4]
```diff
@@ -76,8 +66,8 @@ public Version(final String version) {
    * Returns the exact and full version number.
    * For example, if we specify:
    * <code>
-   * new Version(1.2.3.4.5).getMicro == 4; // true
-   * new Version(1.2.3.4.5).toString().equals("1.2.3.4"); // true
+   * new Version(1.2.3).getPoint == 3; // true
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/util/Version.java]: I think the purpose of this comment was to demonstrate that anything after the last component was ignored by the parser and would not affect the component accessors, so using the same version string as the other examples still seems appropriate:

```suggestion
   * new Version(1.2.3.4.5).getPoint == 3; // true
```

Since `micro` has been removed, it might make sense to change all the examples to use `1.2.3.4` as input instead of `1.2.3.4.5`.

File: game-core/src/test/java/games/strategy/util/VersionTest.java [hunk 5]
```diff
@@ -80,28 +77,23 @@ public void testIsLessThan() {
     assertFalse(new Version(2, 0).isLessThan(new Version(1, 0)));
   }
 
-  @Test
-  public void testWithMicro() {
```
Comment (ssoloff) [inline game-core/src/test/java/games/strategy/util/VersionTest.java]: Instead of removing this test, it should probably be replaced with an equivalent test for the new `withPoint()` method.

**Comment by DanVanAtta** [comment]
@simon33-2 this was a lot slower getting over the finish line than I would have hoped, I'm sorry for that. You are really close, this PR is good with me, please work with Ssoloff & address the 3 remaining small comments, and we'll have this completed. 

If you won't have time in the following week or two, let us know and we can likely complete this.

**Comment by simon33-2** [comment]
Had an error connecting to my private lobby. Seems that it can't get the password for my login even though I'm logging in anonymously.

INFO: Lobby started
Dec 06, 2018 7:42:42 PM org.postgresql.core.v3.ConnectionFactoryImpl log
WARNING: ConnectException occurred while connecting to localhost:5432
java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at org.postgresql.core.PGStream.<init>(PGStream.java:69)
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:156)
        at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
        at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)
        at org.postgresql.Driver.makeConnection(Driver.java:452)
        at org.postgresql.Driver.connect(Driver.java:254)
        at java.sql.DriverManager.getConnection(DriverManager.java:664)
        at java.sql.DriverManager.getConnection(DriverManager.java:208)
        at org.triplea.lobby.server.db.Database.newConnection(Database.java:32)
        at org.triplea.lobby.server.db.AbstractController.newDatabaseConnection(AbstractController.java:25)
        at org.triplea.lobby.server.db.UserController.getPassword(UserController.java:36)
        at org.triplea.lobby.server.db.UserController.getLegacyPassword(UserController.java:27)
        at org.triplea.lobby.server.login.LobbyLoginValidator.newMd5CryptAuthenticatorChallenge(LobbyLoginValidator.java:121)
        at org.triplea.lobby.server.login.LobbyLoginValidator.getChallengeProperties(LobbyLoginValidator.java:111)
        at games.strategy.net.nio.ServerQuarantineConversation.message(ServerQuarantineConversation.java:69)
        at games.strategy.net.nio.Decoder.sendQuarantine(Decoder.java:109)
        at games.strategy.net.nio.Decoder.loop(Decoder.java:81)
        at java.lang.Thread.run(Thread.java:748)

Dec 06, 2018 7:42:42 PM org.postgresql.Driver connect
SEVERE: Connection error: 
org.postgresql.util.PSQLException: Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:245)
        at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
        at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)
        at org.postgresql.Driver.makeConnection(Driver.java:452)
        at org.postgresql.Driver.connect(Driver.java:254)
        at java.sql.DriverManager.getConnection(DriverManager.java:664)
        at java.sql.DriverManager.getConnection(DriverManager.java:208)
        at org.triplea.lobby.server.db.Database.newConnection(Database.java:32)
        at org.triplea.lobby.server.db.AbstractController.newDatabaseConnection(AbstractController.java:25)
        at org.triplea.lobby.server.db.UserController.getPassword(UserController.java:36)
        at org.triplea.lobby.server.db.UserController.getLegacyPassword(UserController.java:27)
        at org.triplea.lobby.server.login.LobbyLoginValidator.newMd5CryptAuthenticatorChallenge(LobbyLoginValidator.java:121)
        at org.triplea.lobby.server.login.LobbyLoginValidator.getChallengeProperties(LobbyLoginValidator.java:111)
        at games.strategy.net.nio.ServerQuarantineConversation.message(ServerQuarantineConversation.java:69)
        at games.strategy.net.nio.Decoder.sendQuarantine(Decoder.java:109)
        at games.strategy.net.nio.Decoder.loop(Decoder.java:81)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at org.postgresql.core.PGStream.<init>(PGStream.java:69)
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:156)
        ... 16 more

Dec 06, 2018 7:42:42 PM org.triplea.lobby.server.db.AbstractController newDatabaseException
SEVERE: Error getting password for user: simon33
org.postgresql.util.PSQLException: Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:245)
        at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
        at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)
        at org.postgresql.Driver.makeConnection(Driver.java:452)
        at org.postgresql.Driver.connect(Driver.java:254)
        at java.sql.DriverManager.getConnection(DriverManager.java:664)
        at java.sql.DriverManager.getConnection(DriverManager.java:208)
        at org.triplea.lobby.server.db.Database.newConnection(Database.java:32)
        at org.triplea.lobby.server.db.AbstractController.newDatabaseConnection(AbstractController.java:25)
        at org.triplea.lobby.server.db.UserController.getPassword(UserController.java:36)
        at org.triplea.lobby.server.db.UserController.getLegacyPassword(UserController.java:27)
        at org.triplea.lobby.server.login.LobbyLoginValidator.newMd5CryptAuthenticatorChallenge(LobbyLoginValidator.java:121)
        at org.triplea.lobby.server.login.LobbyLoginValidator.getChallengeProperties(LobbyLoginValidator.java:111)
        at games.strategy.net.nio.ServerQuarantineConversation.message(ServerQuarantineConversation.java:69)
        at games.strategy.net.nio.Decoder.sendQuarantine(Decoder.java:109)
        at games.strategy.net.nio.Decoder.loop(Decoder.java:81)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at org.postgresql.core.PGStream.<init>(PGStream.java:69)
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:156)
        ... 16 more

Dec 06, 2018 7:42:42 PM games.strategy.net.nio.ServerQuarantineConversation message
SEVERE: Error with connection
java.lang.IllegalStateException: Error getting password for user: simon33 (Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.)
        at org.triplea.lobby.server.db.AbstractController.newDatabaseException(AbstractController.java:30)
        at org.triplea.lobby.server.db.UserController.getPassword(UserController.java:47)
        at org.triplea.lobby.server.db.UserController.getLegacyPassword(UserController.java:27)
        at org.triplea.lobby.server.login.LobbyLoginValidator.newMd5CryptAuthenticatorChallenge(LobbyLoginValidator.java:121)
        at org.triplea.lobby.server.login.LobbyLoginValidator.getChallengeProperties(LobbyLoginValidator.java:111)
        at games.strategy.net.nio.ServerQuarantineConversation.message(ServerQuarantineConversation.java:69)
        at games.strategy.net.nio.Decoder.sendQuarantine(Decoder.java:109)
        at games.strategy.net.nio.Decoder.loop(Decoder.java:81)
        at java.lang.Thread.run(Thread.java:748)

**Comment by RoiEXLab** [comment]
@simon33-2 The error you posted indicates your database is down or running on a different port

**Comment by simon33-2** [comment]
Thought I needed to install postgresql but doing so didn't achieve anything.

**Comment by RoiEXLab** [comment]
Well simply installing it isn't enough.
You need to start the database and run the SQL scripts to setup the schema

**Comment by RoiEXLab** [comment]
You can find all of the SQL files here: https://github.com/triplea-game/triplea/tree/master/lobby-db/src/main/resources/db/migration

**Comment by simon33-2** [comment]
Ok, I've run the scripts but I got an error at one point of V1.01:
WARNING:  database "ta_users" does not exist

Still doesn't work. Do I need to do something to give my userid access to the database?

**Comment by RoiEXLab** [comment]
Ah sorry, forgot to mention this:
There's a difference between database system (I.e. Postgres) and the actual database "ta_users". This one needs to be created once before performing all the operations. Not sure what exact tools you're using to manipulate, but Google should be able to help you with that.

**Comment by simon33-2** [comment]
Yep, I understand that part.

I've created the schema for both the postgres user/db and the simon user/db and it seems to not be enough.

Does there need to be a password set for the postgres account, perhaps at the UNIX level or is it at the DB level? Does the password need to be something specific?

**Comment by RoiEXLab** [comment]
You can configure username and password in this file: https://github.com/triplea-game/triplea/blob/master/lobby/config/lobby/lobby.properties

**Comment by simon33-2** [comment]
BTW, the error comes from this line in V1.01:
comment on database ta_users is 'The Database of the TripleA Lobby';

Unless I'm supposed to create a ta_users database and put the tables in that?

**Comment by RoiEXLab** [comment]
Yeah, the db assumes the database to be called ta_users AFAIK not to be confused with the table with the same name.

**Comment by ssoloff** [comment]
> Thought I needed to install postgresql but doing so didn't achieve anything.

@simon33-2 For future reference...  We have a complete Docker image for the lobby database that Just Works.  IIRC, you're running Linux, so your distro will either already have Docker installed or it's just one package manager install away. :smile: See https://github.com/triplea-game/triplea/tree/master/lobby-db#docker for more information.

**Comment by ssoloff** [comment]
> Thought I needed to install postgresql but doing so didn't achieve anything.

@simon33-2 For future reference...  We have a complete Docker image for the lobby database that Just Works.  IIRC, you're running Linux, so your distro will either already have Docker installed or it's just one package manager install away. :smile: See https://github.com/triplea-game/triplea/tree/master/lobby-db#docker for more information.
